File: mac/DeskLayer/DeskLayer/Core/PluginUpdater.swift
//
//  PluginUpdater.swift
//  DeskLayer
//
//  Checks a plugin's updateURL for a newer version and installs it by
//  overwriting the plugin file. Per-plugin auto-update is a user preference;
//  when on, DeskLayer checks at launch (and on demand).
//

import Foundation
import os

nonisolated enum UpdateResult: Sendable, Equatable {
    case upToDate(version: String)
    case updated(from: String, to: String)
    case noUpdateURL
    case failed(String)

    var message: String {
        switch self {
        case .upToDate(let v): return "Up to date (\(v))"
        case .updated(let from, let to): return "Updated \(from) → \(to)"
        case .noUpdateURL: return "No update URL declared"
        case .failed(let why): return "Update failed: \(why)"
        }
    }
}

@MainActor
final class PluginUpdater {
    private let log = Logger(subsystem: "com.qiudaomao.DeskLayer", category: "updater")
    private let session: URLSession = {
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 20
        config.urlCache = nil
        config.requestCachePolicy = .reloadIgnoringLocalCacheData
        return URLSession(configuration: config)
    }()

    /// Never served from cache: raw.githubusercontent.com sends
    /// Cache-Control: max-age=300, so a user who edits a plugin and checks for
    /// an update would keep seeing the old one for five minutes.
    private func request(_ url: URL) -> URLRequest {
        var request = URLRequest(url: url)
        request.cachePolicy = .reloadIgnoringLocalCacheData
        return request
    }


    // Per-plugin "auto-update" preference, persisted in UserDefaults.
    private static let autoKey = "DeskLayer.autoUpdatePlugins"

    func isAutoUpdate(_ pluginID: String) -> Bool {
        autoSet().contains(pluginID)
    }

    func setAutoUpdate(_ on: Bool, for pluginID: String) {
        var set = autoSet()
        if on { set.insert(pluginID) } else { set.remove(pluginID) }
        UserDefaults.standard.set(Array(set), forKey: Self.autoKey)
    }

    private func autoSet() -> Set<String> {
        Set(UserDefaults.standard.stringArray(forKey: Self.autoKey) ?? [])
    }

    /// Update check. `installedSource` is the current file (to read its
    /// version and updateURL); `destination` is where a newer version writes.
    ///
    /// Efficient path: fetch a small sibling manifest — same path/name with a
    /// `.json` extension (Clock.js → Clock.json) — holding {version, url}, and
    /// download the plugin body only when it's actually newer. If no manifest
    /// exists, fall back to fetching the `.js` at updateURL directly.
    func check(
        pluginID: String,
        installedSource: String,
        destination: URL
    ) async -> UpdateResult {
        let localMeta = PluginMetadata.extract(from: installedSource)
        guard let urlString = localMeta.updateURL, let updateURL = URL(string: urlString), updateURL.scheme != nil else {
            return .noUpdateURL
        }
        let localVersion = localMeta.version ?? "0"

        // 1) Try the manifest (small JSON) first.
        if let manifest = await fetchManifest(for: updateURL) {
            guard compareVersions(manifest.version, localVersion) == .orderedDescending else {
                return .upToDate(version: localVersion)
            }
            let bodyURL = manifest.url.flatMap { URL(string: $0) } ?? bodyURL(for: updateURL)
            return await download(bodyURL, into: destination, from: localVersion, to: manifest.version, pluginID: pluginID)
        }

        // 2) Fallback: fetch the .js directly and read its declared version.
        do {
            let (data, response) = try await session.data(for: request(updateURL))
            if let http = response as? HTTPURLResponse, !(200...299).contains(http.statusCode) {
                return .failed("HTTP \(http.statusCode)")
            }
            guard let remoteSource = String(data: data, encoding: .utf8), !remoteSource.isEmpty else {
                return .failed("response was not text")
            }
            let remoteVersion = PluginMetadata.extract(from: remoteSource).version ?? "0"
            guard compareVersions(remoteVersion, localVersion) == .orderedDescending else {
                return .upToDate(version: localVersion)
            }
            try data.write(to: destination, options: .atomic)
            log.info("updated \(pluginID, privacy: .public) \(localVersion, privacy: .public) → \(remoteVersion, privacy: .public)")
            return .updated(from: localVersion, to: remoteVersion)
        } catch {
            return .failed(error.localizedDescription)
        }
    }

    private struct Manifest: Decodable {
        let version: String
        let url: String?
    }

    /// The manifest URL is the updateURL with its extension swapped to .json
    /// (or updateURL itself when it already points at a .json).
    private func manifestURL(for updateURL: URL) -> URL {
        if updateURL.pathExtension.lowercased() == "json" { return updateURL }
        return updateURL.deletingPathExtension().appendingPathExtension("json")
    }

    /// Where to download the plugin body when a manifest omits an explicit url:
    /// the updateURL itself, unless it's the .json manifest, in which case .js.
    private func bodyURL(for updateURL: URL) -> URL {
        if updateURL.pathExtension.lowercased() == "json" {
            return updateURL.deletingPathExtension().appendingPathExtension("js")
        }
        return updateURL
    }

    private func fetchManifest(for updateURL: URL) async -> Manifest? {
        let url = manifestURL(for: updateURL)
        guard let (data, response) = try? await session.data(for: request(url)) else { return nil }
        if let http = response as? HTTPURLResponse, !(200...299).contains(http.statusCode) { return nil }
        return try? JSONDecoder().decode(Manifest.self, from: data)
    }

    private func download(
        _ url: URL, into destination: URL, from localVersion: String, to remoteVersion: String, pluginID: String
    ) async -> UpdateResult {
        do {
            let (data, response) = try await session.data(for: request(url))
            if let http = response as? HTTPURLResponse, !(200...299).contains(http.statusCode) {
                return .failed("HTTP \(http.statusCode) fetching plugin body")
            }
            guard let source = String(data: data, encoding: .utf8), !source.isEmpty else {
                return .failed("plugin body was not text")
            }
            try data.write(to: destination, options: .atomic)
            log.info("updated \(pluginID, privacy: .public) \(localVersion, privacy: .public) → \(remoteVersion, privacy: .public) (manifest)")
            return .updated(from: localVersion, to: remoteVersion)
        } catch {
            return .failed(error.localizedDescription)
        }
    }
}

File: win/src/DeskLayer.Core/PluginUpdater.cs
// Per-plugin updater — port of the mac PluginUpdater.swift. Reads a plugin's
// declared updateURL + version, fetches a small sibling manifest (Clock.js →
// Clock.json holding {version, url}), and overwrites the .js only when the
// remote version is newer. Falls back to fetching the .js directly. Per-plugin
// auto-update is a preference (auto-update.json in the data dir).

using System.Text.Json;
using DeskLayer.Core.Model;

namespace DeskLayer.Core;

public enum UpdateOutcome { UpToDate, Updated, NoUpdateUrl, Failed }

public readonly record struct UpdateResult(UpdateOutcome Outcome, string Message);

public sealed class PluginUpdater
{
    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(20) };
    private static string AutoFile => Path.Combine(LayoutStore.DataDirectory, "auto-update.json");

    private readonly Action<string> log;
    public PluginUpdater(Action<string> log) => this.log = log;

    // ---- auto-update preference ----

    public bool IsAutoUpdate(string pluginId) => LoadAuto().Contains(pluginId);

    public void SetAutoUpdate(string pluginId, bool on)
    {
        var set = LoadAuto();
        if (on) set.Add(pluginId); else set.Remove(pluginId);
        try { File.WriteAllText(AutoFile, JsonSerializer.Serialize(set)); } catch { }
    }

    private HashSet<string> LoadAuto()
    {
        try
        {
            return File.Exists(AutoFile)
                ? JsonSerializer.Deserialize<HashSet<string>>(File.ReadAllText(AutoFile)) ?? new()
                : new();
        }
        catch { return new(); }
    }

    // ---- update check ----

    /// `installedSource` is read for version + updateURL; `destination` is
    /// where a newer body writes.
    public async Task<UpdateResult> Check(string pluginId, string installedSource, string destination)
    {
        var (localVersion, updateUrlString) = PluginMetadata.Extract(installedSource);
        if (string.IsNullOrEmpty(updateUrlString) || !Uri.TryCreate(updateUrlString, UriKind.Absolute, out var updateUrl))
            return new(UpdateOutcome.NoUpdateUrl, "No update URL declared");
        localVersion ??= "0";

        // 1) Manifest first (small JSON).
        var manifest = await FetchManifest(updateUrl);
        if (manifest != null)
        {
            if (CompareVersions(manifest.Version, localVersion) <= 0)
                return new(UpdateOutcome.UpToDate, $"Up to date ({localVersion})");
            var bodyUrl = !string.IsNullOrEmpty(manifest.Url) && Uri.TryCreate(manifest.Url, UriKind.Absolute, out var u)
                ? u : BodyUrl(updateUrl);
            return await Download(bodyUrl, destination, localVersion, manifest.Version, pluginId);
        }

        // 2) Fallback: fetch the .js and read its declared version.
        try
        {
            using var req = new HttpRequestMessage(HttpMethod.Get, updateUrl);
            req.Headers.CacheControl = new() { NoCache = true };
            var response = await Http.SendAsync(req);
            if (!response.IsSuccessStatusCode) return new(UpdateOutcome.Failed, $"Update failed: HTTP {(int)response.StatusCode}");
            var remoteSource = await response.Content.ReadAsStringAsync();
            if (string.IsNullOrEmpty(remoteSource)) return new(UpdateOutcome.Failed, "Update failed: response was not text");
            var (remoteVersion, _) = PluginMetadata.Extract(remoteSource);
            remoteVersion ??= "0";
            if (CompareVersions(remoteVersion, localVersion) <= 0)
                return new(UpdateOutcome.UpToDate, $"Up to date ({localVersion})");
            File.WriteAllText(destination, remoteSource);
            log($"updated {pluginId} {localVersion} → {remoteVersion}");
            return new(UpdateOutcome.Updated, $"Updated {localVersion} → {remoteVersion}");
        }
        catch (Exception ex) { return new(UpdateOutcome.Failed, $"Update failed: {ex.Message}"); }
    }

    private sealed class Manifest { public string Version { get; set; } = "0"; public string? Url { get; set; } }

    private static Uri ManifestUrl(Uri updateUrl) =>
        updateUrl.AbsolutePath.EndsWith(".json", StringComparison.OrdinalIgnoreCase)
            ? updateUrl
            : new Uri(updateUrl.AbsoluteUri[..^Path.GetExtension(updateUrl.AbsolutePath).Length] + ".json");

    private static Uri BodyUrl(Uri updateUrl) =>
        updateUrl.AbsolutePath.EndsWith(".json", StringComparison.OrdinalIgnoreCase)
            ? new Uri(updateUrl.AbsoluteUri[..^5] + ".js")
            : updateUrl;

    private async Task<Manifest?> FetchManifest(Uri updateUrl)
    {
        try
        {
            using var req = new HttpRequestMessage(HttpMethod.Get, ManifestUrl(updateUrl));
            req.Headers.CacheControl = new() { NoCache = true };
            var response = await Http.SendAsync(req);
            if (!response.IsSuccessStatusCode) return null;
            return JsonSerializer.Deserialize<Manifest>(await response.Content.ReadAsStringAsync(),
                new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
        }
        catch { return null; }
    }

    private async Task<UpdateResult> Download(Uri url, string destination, string from, string to, string pluginId)
    {
        try
        {
            using var req = new HttpRequestMessage(HttpMethod.Get, url);
            req.Headers.CacheControl = new() { NoCache = true };
            var response = await Http.SendAsync(req);
            if (!response.IsSuccessStatusCode) return new(UpdateOutcome.Failed, $"Update failed: HTTP {(int)response.StatusCode}");
            var source = await response.Content.ReadAsStringAsync();
            if (string.IsNullOrEmpty(source)) return new(UpdateOutcome.Failed, "Update failed: plugin body was not text");
            File.WriteAllText(destination, source);
            log($"updated {pluginId} {from} → {to} (manifest)");
            return new(UpdateOutcome.Updated, $"Updated {from} → {to}");
        }
        catch (Exception ex) { return new(UpdateOutcome.Failed, $"Update failed: {ex.Message}"); }
    }

    /// Dotted-numeric compare (1.2.10 > 1.2.9). Missing components are 0.
    public static int CompareVersions(string a, string b)
    {
        var pa = a.Split('.'); var pb = b.Split('.');
        for (var i = 0; i < Math.Max(pa.Length, pb.Length); i++)
        {
            var na = i < pa.Length && int.TryParse(pa[i], out var x) ? x : 0;
            var nb = i < pb.Length && int.TryParse(pb[i], out var y) ? y : 0;
            if (na != nb) return na.CompareTo(nb);
        }
        return 0;
    }
}
